test/run_test.py

```diff
@@ -677,6 +704,60 @@ def run_doctests(test_module, test_directory, options):
     return result
 
 
+def print_log_file(test, file_path):
+    with open(file_path, "r") as f:
+        print_to_stderr("")
+        print_to_stderr(f'PRINT LOG FILE of {test} ({file_path})')
+        print_to_stderr(f'##[group]PRINT LOG FILE of {test} ({file_path})')
+        print_to_stderr(f.read())
+        print_to_stderr('##[endgroup]')
+        print_to_stderr(f"FINISHED PRINT LOG FILE of {test} ({file_path})")
+        print_to_stderr("")
+    os.remove(file_path)
+
+
+def run_test_ops(test_module, test_directory, options):
+    if 'slow-gradcheck' in os.getenv("BUILD_ENVIRONMENT", ""):
+        # there are a lot of tests that take up a lot of space in slowgrad check, so don't bother parallelizing
+        # it's also on periodic so we don't care about TTS as much
+        return run_test(test_module, test_directory, copy.deepcopy(options),
+                        extra_unittest_args=["--use-pytest", '-vv', '-x', '--reruns=2', '-rfEX'],
+                        )
+
+    file_names = []
+    return_codes = []
+    num_procs = 3
```

Comment:
Any reason behind this number of process?  I think we could try to use the number of CPU here instead.  That's what people commonly use, but we need to try it out to see if it works better or now

Reply:
when i did testing with pytest-xdist, setting -n=auto, which i think uses the number of CPUs, mac had 3 CPUs and also received the best speedup with 3 processes.  I'm also a little hesitant to increase it beyond 3 or 4 because then we would have to restrict the cuda memory more

Reply:
I see. Sound good then, we can make this a constant, i.e. NUM_PROCS, then, so people know what to expect.  You could just comment exactly what you mention here :)